### PR TITLE
drop chromium-pickle dependency. implement reading / writing asar archives in pure JS

### PR DESCRIPTION
**Pull request by YuriSolovyov**
As follow-up to https://github.com/atom/asar/issues/16

### CODE REVIEW

**Comment by YuriSolovyov** [comment]
investigating failure

**Comment by YuriSolovyov** [comment]
works even with 0.10 now, not sure if it is needed though. /cc @zcbenz

**Comment by zcbenz** [comment]
Can you maybe move the operations of reading/writing to a new class or module? In the original code it is plain to see that we have written two Pickle objects into file, but in this PR they become confusing operations of `Buffer` and magic numbers.

File: lib/header.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    throw new Error('Unable to read header');
+
+  return headerBuf.toString('utf8');
+};
+
+var writeHeader = function(dest, headerBuffer, sizeBuffer, callback) {
+  dest.write(pickleHeader);
+  dest.write(sizeBuffer);
+  dest.write(headerBuffer);
+  dest.write(headerTerminator, callback);
```
Comment (YuriSolovyov) [inline lib/header.js]: maybe it is better to do
```javascript
  var finalBuffer = Buffer.concat([
      pickleHeader,
      sizeBuffer,
      headerBuffer,
      headerTerminator
  ]);
  dest.write(finalBuffer, callback);
```
here?

**Comment by YuriSolovyov** [comment]
moved related code to own file, just have one question.

File: lib/header.js (hunk 1 above)
Comment (bwin) [inline lib/header.js]: IMHO this `Buffer.concat` just makes an unnecessary copy.

**Comment by zcbenz** [comment]
It looks good to me, thanks!

**Comment by zcbenz** [comment]
asar@0.3.0 has been published for this change.

**Comment by YuriSolovyov** [comment]
@zcbenz can you confirm it works on windows?
It fails to launch an app built with new version for me, and I can't even investigate it till morning.

**Comment by YuriSolovyov** [comment]
@zcbenz even though test pass, it does not work on linux either. shell pickle fails to read header string
```
user@y:~/fix-asar$ node_modules/.bin/asar l app.asar
/home/user/fix-asar/node_modules/asar/lib/disk.js:47
  var header = headerPickle.createIterator().readString();
                                             ^
TypeError: Data can not be extracted
    at TypeError (native)
    at Object.module.exports.readArchiveHeader (/home/user/fix-asar/node_modules/asar/lib/disk.js:47:46)
    at Object.module.exports.readFilesystem (/home/user/fix-asar/node_modules/asar/lib/disk.js:52:21)
    at Object.module.exports.listPackage (/home/user/fix-asar/node_modules/asar/lib/asar.js:67:15)
    at Command.<anonymous> (/home/user/fix-asar/node_modules/asar/bin/asar:17:23)
    at Command.listener (/home/user/fix-asar/node_modules/asar/node_modules/commander/index.js:249:8)
    at Command.emit (events.js:110:17)
    at Command.parseArgs (/home/user/fix-asar/node_modules/asar/node_modules/commander/index.js:480:12)
    at Command.parse (/home/user/fix-asar/node_modules/asar/node_modules/commander/index.js:372:21)
    at Object.<anonymous> (/home/user/fix-asar/node_modules/asar/bin/asar:37:9)

```
my bad. I should've tried it on simple app, and not rely on test passing only.

**Comment by YuriSolovyov** [comment]
looks like it might be something with hardcoded pickle header
```
user@y:~/fix-asar$ xxd app.asar > b1.hex
user@y:~/fix-asar$ xxd app-new.asar > b2.hex
user@y:~/fix-asar$ diff b1.hex b2.hex
1c1
< 0000000: 0400 0000 d00a 0000 cc0a 0000 c60a 0000  ................
---
> 0000000: 0400 0000 fc00 0000 f800 0000 c60a 0000  ................
```

**Pull request merged**